Ical.Net/Evaluation/EventEvaluator.cs

```diff
@@ -27,37 +31,69 @@ public EventEvaluator(CalendarEvent evt) : base(evt) { }
     /// <param name="periodEnd">The end date of the range to evaluate.</param>
     /// <param name="includeReferenceDateInResults"></param>
     /// <returns></returns>
     public override IEnumerable<Period> Evaluate(IDateTime referenceTime, DateTime? periodStart, DateTime? periodEnd, bool includeReferenceDateInResults)
     {
-        Period WithDuration(Period period)
-        {
-            var duration = CalendarEvent.GetFirstDuration();
-            var endTime = duration == default
-                ? period.StartTime
-                : period.StartTime.Add(CalendarEvent.GetFirstDuration());
-
-            return new Period(period.StartTime)
-            {
-                Duration = duration,
-                EndTime = endTime,
-            };
-        }
-
         // Evaluate recurrences normally
         var periods = base.Evaluate(referenceTime, periodStart, periodEnd, includeReferenceDateInResults)
-            .Select(WithDuration);
+            .Select(WithEndTime);
 
         return periods;
     }
+
+    /// <summary>
+    /// The <paramref name="period"/> to evaluate has the <see cref="Period.StartTime"/> set,
+    /// but neither <see cref="Period.EndTime"/> nor <see cref="Period.Duration"/> are set.
+    /// </summary>
+    /// <param name="period">The period where <see cref="Period.EndTime"/> will be set.</param>
+    /// <returns>Returns the <paramref name="period"/> with <see cref="Period.EndTime"/> and exact <see cref="Period.Duration"/> set.</returns>
+    private Period WithEndTime(Period period)
+    {
+        /*
+           Calculate the nominal duration of the event which is the duration of the
+           first occurrence from the event's definition.
+           
+           The nominal duration is used, because the period end time gets the same timezone as the event end time.
+           This ensures that the end time is correct, even for DST transitions.
+           The exact duration is calculated from the zoned end time and the zoned start time,
+           and it may differ from the nominal duration.
+         */
+        var nominalDuration = CalendarEvent.CalcFirstNominalDuration();
+
+        IDateTime endTime;
+        if (nominalDuration == TimeSpan.Zero)
+        {
+            // For a zero-duration event, the end time is the same as the start time.
+            endTime = period.StartTime;
+        }
+        else
+        {
+            // Calculate the end time of the event as a DateTime
+            var endDt = period.StartTime.Value.Add(nominalDuration);
+
+            // Create a CalDateTime object with the calculated end time.
+            // Ensure date-only or date-time consistency with the start time,
+            // and use the timezone from the event's definition.
+            endTime = new CalDateTime(
+                DateOnly.FromDateTime(endDt),
+                period.StartTime.HasTime ? TimeOnly.FromDateTime(endDt) : null,
```

Comment:
Here we risk confusing the tzid. Simply use `var endDt = period.StartTime.Add(nominalDuration)`? We must make sure that Add() and GetFirst..() both operate on the same kind of duration, nominal vs exact.

Reply:
See [comment](https://github.com/ical-org/ical.net/pull/673#issuecomment-2558053226)